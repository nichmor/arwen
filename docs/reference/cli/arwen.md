<!--- This file is autogenerated. Do not edit manually! -->
# <code>arwen</code>

## About
The `arwen`

--8<-- "docs/reference/cli/arwen_extender.md:description"

## Usage
```
arwen <COMMAND>
```

## Subcommands
| Command | Description |
|---------|-------------|
| [`macho`](arwen/macho.md) | Mach-O commands |
| [`elf`](arwen/elf.md) | ELF commands |


--8<-- "docs/reference/cli/arwen_extender.md:example"
