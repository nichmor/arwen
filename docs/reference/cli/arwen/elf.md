<!--- This file is autogenerated. Do not edit manually! -->
# <code>[arwen](../arwen.md) elf</code>

## About
ELF commands

--8<-- "docs/reference/cli/arwen/elf_extender.md:description"

## Usage
```
arwen elf <COMMAND>
```

## Subcommands
| Command | Description |
|---------|-------------|
| [`add-rpath`](elf/add-rpath.md) | Add a run path to the elf file |
| [`remove-rpath`](elf/remove-rpath.md) | Remove runpath from the elf file |
| [`set-rpath`](elf/set-rpath.md) | Set the rpath of the elf file |
| [`print-rpath`](elf/print-rpath.md) | Print the rpath of the ELF file |
| [`force-rpath`](elf/force-rpath.md) | Force usage of `DT_RPATH` instead of `DT_RUNPATH` |
| [`set-interpreter`](elf/set-interpreter.md) | Set the interpreter of the elf file |
| [`print-interpreter`](elf/print-interpreter.md) | Print the interpreter of the ELF file |
| [`set-os-abi`](elf/set-os-abi.md) | Change the OS ABI of the elf file |
| [`print-os-abi`](elf/print-os-abi.md) | Print the OS ABI of the elf file |
| [`set-soname`](elf/set-soname.md) | Sets the DT_SONAME field in the ELF file |
| [`print-soname`](elf/print-soname.md) | Print the `DT_SONAME` |
| [`shrink-rpath`](elf/shrink-rpath.md) | Remove from the DT_RUNPATH or DT_RPATH all directories that do not contain a library referenced by DT_NEEDED fields of the executable or library |
| [`add-needed`](elf/add-needed.md) | Add dependencies on a dynamic library  DT_NEEDED |
| [`remove-needed`](elf/remove-needed.md) | Remove dynamic library dependencies from DT_NEEDED |
| [`replace-needed`](elf/replace-needed.md) | Replace dependencies from DT_NEEDED |
| [`print-needed`](elf/print-needed.md) | Print the DT_NEEDED |
| [`no-default-lib`](elf/no-default-lib.md) | Disable the default library search paths in the elf file |
| [`clear-symbol-version`](elf/clear-symbol-version.md) | Clear a version for a specific symbol |
| [`rename-dynamic-symbols`](elf/rename-dynamic-symbols.md) | Renames dynamic symbols |
| [`add-debug-tag`](elf/add-debug-tag.md) | Adds `DT_DEBUG` tag to the elf file to enable debugging |
| [`clear-exec-stack`](elf/clear-exec-stack.md) | Clears the executable flag of the GNU_STACK program header |
| [`set-exec-stack`](elf/set-exec-stack.md) | Set the executable flag of the GNU_STACK program header |
| [`print-exec-stack`](elf/print-exec-stack.md) | Print the exec stack of the ELF file |


--8<-- "docs/reference/cli/arwen/elf_extender.md:example"
