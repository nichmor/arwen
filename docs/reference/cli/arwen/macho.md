<!--- This file is autogenerated. Do not edit manually! -->
# <code>[arwen](../arwen.md) macho</code>

## About
Mach-O commands

--8<-- "docs/reference/cli/arwen/macho_extender.md:description"

## Usage
```
arwen macho <COMMAND>
```

## Subcommands
| Command | Description |
|---------|-------------|
| [`delete-rpath`](macho/delete-rpath.md) | Remove a run path |
| [`change-rpath`](macho/change-rpath.md) | Change already existing run path |
| [`add-rpath`](macho/add-rpath.md) | Add a run path |
| [`change-install-name`](macho/change-install-name.md) | Change existing dylib load name |
| [`change-install-id`](macho/change-install-id.md) | Change dylib id. Works only if your object is a shared library |


--8<-- "docs/reference/cli/arwen/macho_extender.md:example"
