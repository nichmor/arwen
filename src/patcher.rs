use std::vec;

use goblin::mach::load_command::{CommandVariant::*, DylibCommand};
use goblin::mach::{
    header::Header64,
    load_command::{self, LoadCommand, RpathCommand},
    MachO,
};


use scroll::Pwrite;

use crate::commands::{DlibCommandBuilder, RpathCommandBuilder};

/// Change the rpath of a Mach-O file.
pub fn change_rpath(bytes_of_file: Vec<u8>, old_rpath: String, new_rpath: String) -> Vec<u8> {
    let parsed_macho = MachO::parse(&bytes_of_file, 0).unwrap();

    // TODO: goblin doesn't export parsed data
    // so we mimic it by using directly the bytes of the file
    // let mut buffer = parsed_macho.data.to_vec();
    let mut buffer = bytes_of_file.to_vec();

    let header = parsed_macho.header;

    let mut header: Header64 = header.into();

    // let's start with load commands
    // initial_offset += wroted;
    let old_rpath_index = parsed_macho
        .rpaths
        .iter()
        .position(|rpath| rpath == &old_rpath)
        .expect("rpath should exist");

    // now based on the index, we need to find the RpathCommand from the load commands
    let (load_command, _rpath_command) =
        find_rpath_command(&parsed_macho.load_commands, old_rpath_index)
            .expect("rpath command should exist");

    remove_load_command(&mut buffer, &mut header, load_command);

    let (new_rpath, new_rpath_command_buffer) =
        RpathCommandBuilder::new(new_rpath.as_str()).build();

    insert_command(
        &mut buffer,
        &mut header,
        load_command.offset,
        new_rpath.cmdsize,
        new_rpath_command_buffer,
    );

    buffer.to_vec()
}

/// Add a new rpath to a Mach-O file.
pub fn add_rpath(bytes_of_file: Vec<u8>, new_rpath: String) -> Vec<u8> {
    // let's calculate the total size of all the header and commands
    let parsed_macho = MachO::parse(&bytes_of_file, 0).unwrap();

    // TODO: goblin doesn't export parsed data
    // so we mimic it by using directly the bytes of the file
    // let mut buffer = parsed_macho.data.to_vec();
    let mut buffer = bytes_of_file.to_vec();

    let header = parsed_macho.header;

    let mut header: Header64 = header.into();

    let (new_rpath, new_rpath_command_buffer) =
        RpathCommandBuilder::new(new_rpath.as_str()).build();

    let offset_size = header.size() + header.sizeofcmds as usize;

    insert_command(
        &mut buffer,
        &mut header,
        offset_size,
        new_rpath.cmdsize,
        new_rpath_command_buffer,
    );

    buffer.to_vec()
}

/// Remove a specified rpath from a Mach-O file.
pub fn remove_rpath(bytes_of_file: Vec<u8>, old_rpath: String) -> Vec<u8> {
    let parsed_macho = MachO::parse(&bytes_of_file, 0).unwrap();

    // TODO: goblin doesn't export parsed data
    // so we mimic it by using directly the bytes of the file
    // let mut buffer = parsed_macho.data.to_vec();
    let mut buffer = bytes_of_file.to_vec();

    let header = parsed_macho.header;

    let mut header: Header64 = header.into();

    let old_rpath_index = parsed_macho
        .rpaths
        .iter()
        .position(|rpath| rpath == &old_rpath)
        .expect("rpath should exist");

    // now based on the index, we need to find the RpathCommand from the load commands
    let (load_command, _rpath_command) =
        find_rpath_command(&parsed_macho.load_commands, old_rpath_index)
            .expect("rpath command should exist");

    remove_load_command(&mut buffer, &mut header, load_command);

    buffer.to_vec()
}

/// Change the install name of a dylib in a Mach-O file.
pub fn change_install_name(bytes_of_file: Vec<u8>, old_name: String, new_name: String) -> Vec<u8> {
    let parsed_macho = MachO::parse(&bytes_of_file, 0).unwrap();

    // TODO: goblin doesn't export parsed data
    // so we mimic it by using directly the bytes of the file
    // let mut buffer = parsed_macho.data.to_vec();
    let mut buffer = bytes_of_file.to_vec();

    let header = parsed_macho.header;

    let mut header: Header64 = header.into();

    // let's start with load commands
    let old_dylib = parsed_macho
        .libs
        .iter()
        .position(|name| name == &old_name)
        .expect("dylib name should exist");

    // now based on the index, we need to find the RpathCommand from the load commands
    // we use -1 because dylib contains self, so we need to omit it
    let (load_command, old_dylib) = find_dylib_command(&parsed_macho.load_commands, old_dylib - 1)
        .expect("rpath command should exist");

    remove_load_command(&mut buffer, &mut header, load_command);

    let (new_dylib, new_dylib_command_buffer) =
        DlibCommandBuilder::new(&new_name, *old_dylib).build();

    insert_command(
        &mut buffer,
        &mut header,
        load_command.offset,
        new_dylib.cmdsize,
        new_dylib_command_buffer,
    );

    buffer.to_vec()
}

/// Change the install id of a dylib in a Mach-O file.
pub fn change_install_id(bytes_of_file: Vec<u8>, new_id: String) -> Vec<u8> {
    let parsed_macho = MachO::parse(&bytes_of_file, 0).unwrap();

    // TODO: goblin doesn't export parsed data
    // so we mimic it by using directly the bytes of the file
    // let mut buffer = parsed_macho.data.to_vec();
    let mut buffer = bytes_of_file.to_vec();

    let header = parsed_macho.header;

    let mut header: Header64 = header.into();

    // now based on the index, we need to find the RpathCommand from the load commands
    let (load_command, old_dylib) = find_dylib_id(&parsed_macho.load_commands)
        .expect("LC_ID_DYLIB is missing or file is not a shared library");

    remove_load_command(&mut buffer, &mut header, load_command);

    let (new_dylib, new_dylib_command_buffer) =
        DlibCommandBuilder::new(&new_id, *old_dylib).build();

    insert_command(
        &mut buffer,
        &mut header,
        load_command.offset,
        new_dylib.cmdsize,
        new_dylib_command_buffer,
    );

    buffer.to_vec()
}

/// Removes a load command from the buffer.
///
/// # Arguments
/// * `buffer` - Mutable byte buffer representing the Mach-O file.
/// * `header` - Header of the macho. It will be updated after removing the load command.
/// * `load_command` - Load Command to remove.
pub fn remove_load_command(
    buffer: &mut Vec<u8>,
    header: &mut Header64,
    load_command: &LoadCommand,
) {
    // // remove entire command from the buffer
    eprintln!("buffer before {}", buffer.len());

    let drain_offset = load_command.offset + load_command.command.cmdsize();
    buffer.drain(load_command.offset..drain_offset);

    eprintln!("buffer after {}", buffer.len());

    // update the header
    header.ncmds -= 1;
    header.sizeofcmds -= load_command.command.cmdsize() as u32;

    // Step 3: Insert padding after the remaining load commands
    let padding_offset = header.size() + header.sizeofcmds as usize;
    let padding_size = load_command.command.cmdsize();

    // Ensure there's enough space for the padding
    if padding_offset + padding_size > buffer.len() {
        buffer.resize(padding_offset + padding_size, 0);
    }

    // Write zero bytes as padding
    let mut zeroing_buffer = vec![0u8; padding_size];
    zeroing_buffer.fill(0);
    eprintln!("zeroing buffer {}", zeroing_buffer.len());

    let tail = buffer.split_off(padding_offset);

    // Extend with the new slice
    buffer.extend(&zeroing_buffer);

    // Add back the tail
    buffer.extend(tail);

    buffer.pwrite(*header, 0).unwrap();
}

/// Insert a new load command at the given offset.
///
/// # Arguments
/// * `buffer` - Mutable byte buffer representing the Mach-O file.
/// * `header` - Header of the macho. It will be updated after removing the load command.
/// * `load_command` - Load Command to remove.
pub fn insert_command(
    buffer: &mut Vec<u8>,
    header: &mut Header64,
    offset: usize,
    new_cmd_size: u32,
    load_data: Vec<u8>,
) {
    // update the header
    header.ncmds += 1;
    header.sizeofcmds += new_cmd_size;

    // write new command
    let tail = buffer.split_off(offset);

    // Extend with the new slice
    buffer.extend(&load_data);

    // Add back the tail
    buffer.extend(tail);

    let drain_offset = header.size() + header.sizeofcmds as usize + new_cmd_size as usize;
    buffer.drain(header.size() + header.sizeofcmds as usize..drain_offset);

    buffer.pwrite(*header, 0).unwrap();
}

/// Find the rpath command at the given index.
pub fn find_rpath_command(
    commands: &[load_command::LoadCommand],
    index: usize,
) -> Option<(&LoadCommand, &RpathCommand)> {
    let mut count = 0;

    for command in commands {
        if let Rpath(rpath_command) = &command.command {
            if count == index {
                return Some((command, rpath_command));
            }
            count += 1;
        }
    }

    None
}

/// Find the dylib command at the given index.
pub fn find_dylib_command(
    commands: &[load_command::LoadCommand],
    index: usize,
) -> Option<(&LoadCommand, &DylibCommand)> {
    let mut count = 0;

    for command in commands {
        eprintln!("command {:?}", command);
        match &command.command {
            LoadDylib(dylib_command)
            | LoadUpwardDylib(dylib_command)
            | ReexportDylib(dylib_command)
            | LoadWeakDylib(dylib_command)
            | LazyLoadDylib(dylib_command) => {
                if count == index {
                    return Some((command, dylib_command));
                }
                count += 1;
            }
            _ => {}
        }
    }

    None
}

/// Find the dylib id command.
pub fn find_dylib_id(
    commands: &[load_command::LoadCommand],
) -> Option<(&LoadCommand, &DylibCommand)> {
    for command in commands {
        if let IdDylib(id_dylib) = &command.command {
            return Some((command, id_dylib));
        }
    }

    None
}
