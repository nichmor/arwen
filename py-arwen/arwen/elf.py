"""
ELF binary manipulation module for Linux/Unix systems.

This module provides functionality for working with ELF binary files,
allowing operations like modifying runpaths, changing interpreters, and more.
"""

from typing import Optional, Dict, List


from arwen.arwen import PyElfContainer


class ElfContainer:
    """
    Python wrapper for manipulating ELF binary files.

    This class provides methods to modify various aspects of ELF binaries,
    such as runpaths, needed libraries, interpreter paths, and more.
    """

    _inner: PyElfContainer

    @staticmethod
    def from_path(path: str) -> "ElfContainer":
        """
        Create a new ElfContainer by parsing an ELF binary file.

        Args:
            path: Path to the ELF binary file

        Returns:
            A new ElfContainer instance

        Raises:
            ElfError: If the file cannot be parsed as an ELF binary
            Exception: If the file cannot be read
        """
        return ElfContainer._from_py(PyElfContainer.from_path(path))

    def add_runpath(self, new_runpath: str) -> None:
        """
        Add a new runpath to the ELF file.

        Args:
            new_runpath: The runpath to add

        Raises:
            ElfError: If the runpath cannot be added
        """
        self._inner.add_runpath(new_runpath)

    def set_runpath(self, set_runpath: str) -> None:
        """
        Set runpath in the ELF file.

        Args:
            set_runpath: The runpath to set

        Raises:
            ElfError: If the runpath cannot be set
        """
        self._inner.set_runpath(set_runpath)

    def remove_runpath(self) -> None:
        """
        Remove any existing runpath from the ELF file.

        Raises:
            ElfError: If the runpath cannot be removed
        """
        self._inner.remove_runpath()

    def get_runpath(self) -> Optional[str]:
        """
        Get the current runpath from the ELF file.

        Returns:
            The current runpath, or None if no runpath is set
        """
        return self._inner.get_runpath()

    def force_rpath(self) -> None:
        """
        Force the ELF file to use DT_RPATH instead of DT_RUNPATH.

        Raises:
            ElfError: If the operation cannot be performed
        """
        self._inner.force_rpath()

    def shrink_rpath(self, rpath_prefixes: List[str]) -> None:
        """
        Shrink rpath to only include specified prefixes.

        Args:
            rpath_prefixes: List of prefixes to include in the rpath

        Raises:
            ElfError: If the rpath cannot be shrunk
        """
        self._inner.shrink_rpath(rpath_prefixes)

    def add_needed(self, needed_libs: List[str]) -> None:
        """
        Add needed libraries to the ELF file.

        Args:
            needed_libs: List of needed libraries to add

        Raises:
            ElfError: If the needed libraries cannot be added
        """
        self._inner.add_needed(needed_libs)

    def remove_needed(self, needed_libs: List[str]) -> None:
        """
        Remove needed libraries from the ELF file.

        Args:
            needed_libs: List of needed libraries to remove

        Raises:
            ElfError: If the needed libraries cannot be removed
        """
        self._inner.remove_needed(needed_libs)

    def replace_needed(self, needed_mappings: Dict[str, str]) -> None:
        """
        Replace needed libraries in the ELF file.

        Args:
            needed_mappings: Dictionary mapping old library names to new library names

        Raises:
            ElfError: If the needed libraries cannot be replaced
        """
        self._inner.replace_needed(needed_mappings)

    def get_needed(self) -> List[str]:
        """
        Get the list of needed libraries from the ELF file.

        Returns:
            List of needed libraries
        """
        return self._inner.get_needed()

    def set_interpreter(self, new_interpreter: str) -> None:
        """
        Change the interpreter path in the ELF file.

        Args:
            new_interpreter: The new interpreter path

        Raises:
            ElfError: If the interpreter cannot be changed
        """
        self._inner.set_interpreter(new_interpreter)

    def get_interpreter(self) -> Optional[str]:
        """
        Get the current interpreter from the ELF file.

        Returns:
            The current interpreter, or None if no interpreter is set
        """
        return self._inner.get_interpreter()

    def set_os_abi(self, os_abi: str) -> None:
        """
        Set the OS ABI in the ELF file.

        Args:
            os_abi: The OS ABI to set

        Raises:
            ElfError: If the OS ABI cannot be set
        """
        self._inner.set_os_abi(os_abi)

    def get_os_abi(self) -> int:
        """
        Get the OS ABI from the ELF file.

        Returns:
            The OS ABI
        """
        return self._inner.get_os_abi()

    def set_soname(self, soname: str) -> None:
        """
        Set the SONAME of the ELF file.

        Args:
            soname: The SONAME to set

        Raises:
            ElfError: If the SONAME cannot be set
        """
        self._inner.set_soname(soname)

    def get_soname(self) -> Optional[str]:
        """
        Get the SONAME of the ELF file.

        Returns:
            The SONAME, or None if no SONAME is set
        """
        return self._inner.get_soname()

    def no_default_lib(self) -> None:
        """
        Disable the default library search paths.

        Raises:
            ElfError: If the operation cannot be performed
        """
        self._inner.no_default_lib()

    def clear_version_symbol(self, symbol_name: str) -> None:
        """
        Clear a specific symbol version in the ELF file.

        Args:
            symbol_name: The name of the symbol whose version to clear

        Raises:
            ElfError: If the version cannot be cleared
        """
        self._inner.clear_version_symbol(symbol_name)

    def add_debug_tag(self) -> None:
        """
        Add a debug tag to the ELF file.

        Raises:
            ElfError: If the debug tag cannot be added
        """
        self._inner.add_debug_tag()

    def clear_exec_stack(self) -> None:
        """
        Clear the executable stack flag in the ELF file.

        Raises:
            ElfError: If the executable stack flag cannot be cleared
        """
        self._inner.clear_exec_stack()

    def set_exec_stack(self) -> None:
        """
        Set the executable stack flag in the ELF file.

        Raises:
            ElfError: If the executable stack flag cannot be set
        """
        self._inner.set_exec_stack()

    def is_exec_stack(self) -> Optional[bool]:
        """
        Get the executable stack status.

        Returns:
            True if the executable stack flag is set, False otherwise, or None if unknown
        """
        return self._inner.is_exec_stack()

    def rename_dynamic_symbols(self, symbols_map: Dict[str, str]) -> None:
        """
        Rename dynamic symbols in the ELF file.

        Args:
            symbols_map: Dictionary mapping old symbol names to new symbol names

        Raises:
            ElfError: If the symbols cannot be renamed
        """
        self._inner.rename_dynamic_symbols(symbols_map)

    def save(self, path: str) -> None:
        """
        Save the modified ELF file to a path.

        Args:
            path: Path where the modified file will be saved

        Raises:
            Exception: If the file cannot be written
        """
        self._inner.save(path)

    @classmethod
    def _from_py(cls, py_elf: PyElfContainer) -> "ElfContainer":
        elf_container = cls.__new__(cls)
        elf_container._inner = py_elf

        return elf_container
