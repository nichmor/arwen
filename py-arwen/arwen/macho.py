"""
Mach-O binary manipulation module for macOS.

This module provides functionality for working with Mach-O binary files,
allowing operations like modifying rpaths, changing install names, and more.
"""

from arwen.arwen import PyMachoContainer


class MachoContainer:
    """
    Python wrapper for manipulating Mach-O binary files.

    This class provides methods to modify various aspects of Mach-O binaries,
    such as rpaths, install names, and install IDs.
    """

    _inner: PyMachoContainer

    @staticmethod
    def from_path(path: str) -> "MachoContainer":
        """
        Create a new MachoContainer by parsing a Mach-O binary file.

        Args:
            path: Path to the Mach-O binary file

        Returns:
            A new MachoContainer instance

        Raises:
            MachoError: If the file cannot be parsed as a Mach-O binary
            Exception: If the file cannot be read
        """
        return MachoContainer._from_py(PyMachoContainer.from_path(path))

    def add_rpath(self, new_rpath: str) -> None:
        """
        Add a new rpath to the Mach-O file.

        Args:
            new_rpath: The rpath to add

        Raises:
            MachoError: If the rpath cannot be added
        """
        self._inner.add_rpath(new_rpath)

    def change_rpath(self, old_rpath: str, new_rpath: str) -> None:
        """
        Change an existing rpath in the Mach-O file.

        Args:
            old_rpath: The existing rpath to change
            new_rpath: The new rpath value

        Raises:
            RpathMissingError: If the old_rpath doesn't exist
            MachoError: If the rpath cannot be changed
        """
        self._inner.change_rpath(old_rpath, new_rpath)

    def remove_rpath(self, old_rpath: str) -> None:
        """
        Remove an existing rpath from the Mach-O file.

        Args:
            old_rpath: The rpath to remove

        Raises:
            RpathMissingError: If the specified rpath doesn't exist
            MachoError: If the rpath cannot be removed
        """
        self._inner.remove_rpath(old_rpath)

    def change_install_id(self, new_id: str) -> None:
        """
        Change the install ID of the Mach-O file (for shared libraries).

        Args:
            new_id: The new install ID

        Raises:
            DylibIdMissingError: If the file is not a shared library
            MachoError: If the install ID cannot be changed
        """
        self._inner.change_install_id(new_id)

    def change_install_name(self, old_name: str, new_name: str) -> None:
        """
        Change the install name of a dependency in the Mach-O file.

        Args:
            old_name: The existing install name to change
            new_name: The new install name

        Raises:
            DylibNameMissingError: If the old_name doesn't exist
            MachoError: If the install name cannot be changed
        """
        self._inner.change_install_name(old_name, new_name)

    def save(self, path: str) -> None:
        """
        Save the modified Mach-O file to a path.

        Args:
            path: Path where the modified file will be saved

        Raises:
            Exception: If the file cannot be written
        """
        self._inner.save(path)

    @classmethod
    def _from_py(cls, py_macho: PyMachoContainer) -> "MachoContainer":
        macho_container = cls.__new__(cls)
        macho_container._inner = py_macho

        return macho_container
