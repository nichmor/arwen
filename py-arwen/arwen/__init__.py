from arwen.macho import MachoContainer
from arwen.elf import ElfContainer
from arwen.arwen import get_arwen_version as _get_arwen_version


__version__ = _get_arwen_version()


__all__ = [
    "MachoContainer",
    "ElfContainer",
]
