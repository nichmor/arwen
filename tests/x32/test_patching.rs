use arwen::patcher::{
    add_rpath, change_install_id, change_install_name, change_rpath, remove_rpath,
};
use goblin::mach::MachO;
use std::path::PathBuf;

/// This test checks if the rpath of a Mach-O file can be changed.
#[test]
fn test_rpath_change() {
    // load hello_with_rpath from data
    let package_dir = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
    let data = package_dir.join("tests/data/x32/ls");

    let data_bytes = std::fs::read(&data).unwrap();

    // we try to change the rpath of the file to something longer
    let changed = change_rpath(
        data_bytes,
        "path_graf".to_string(),
        "path_graf_path_graf_path_graf_path_graf".to_string(),
    );

    let changed_macho = MachO::parse(&changed, 0).unwrap();

    insta::assert_debug_snapshot!(changed_macho);
    insta::assert_debug_snapshot!(changed_macho.rpaths);

    insta::assert_snapshot!(changed.len(), @"49120");
}

/// This test checks if the rpath of a Mach-O file can be removed.
#[test]
fn test_rpath_remove() {
    // load hello_with_rpath from data
    let package_dir = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
    let data = package_dir.join("tests/data/hello_with_rpath");

    let data_bytes = std::fs::read(&data).unwrap();

    // we try to change the rpath of the file to something longer
    let changed = remove_rpath(data_bytes, "path_graf".to_string());

    let changed_macho = MachO::parse(&changed, 0).unwrap();

    insta::assert_debug_snapshot!(changed_macho);
    insta::assert_debug_snapshot!(changed_macho.rpaths);
    insta::assert_snapshot!(changed.len(), @"33440");
}

/// This test verify that a rpath can be added to a Mach-O file.
#[test]
fn test_rpath_add() {
    // load hello_with_rpath from data
    let package_dir = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
    let data = package_dir.join("tests/data/hello_with_rpath");

    let data_bytes = std::fs::read(&data).unwrap();

    // we try to change the rpath of the file to something longer
    let changed = add_rpath(
        data_bytes,
        "abababababababababababaabbababababababababab".to_string(),
    );

    let changed_macho = MachO::parse(&changed, 0).unwrap();

    insta::assert_debug_snapshot!(changed_macho);
    insta::assert_debug_snapshot!(changed_macho.rpaths);
    insta::assert_snapshot!(changed.len(), @"33440");
}

/// This test verify that an dylib id can be changed in a Mach-O file.
#[test]
fn test_change_dylib_id() {
    // load hello_with_rpath from data
    let package_dir = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
    let data = package_dir.join("tests/data/libmylib.dylib");

    let data_bytes = std::fs::read(&data).unwrap();

    // we try to change the rpath of the file to something longer
    let changed = change_install_id(
        data_bytes,
        "very_very_very_very_very_very_very_longid.dylib".to_string(),
    );

    let changed_macho = MachO::parse(&changed, 0).unwrap();

    insta::assert_debug_snapshot!(changed_macho);
    insta::assert_debug_snapshot!(changed_macho.name);
    insta::assert_snapshot!(changed.len(), @"33355");
}

/// This test verify that a name of dynamic library can be changed
#[test]
fn test_change_dylib_name() {
    // load hello_with_rpath from data
    let package_dir = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
    let data = package_dir.join("tests/data/hello_with_rpath");

    let data_bytes = std::fs::read(&data).unwrap();

    // we try to change the rpath of the file to something longer
    let changed = change_install_name(
        data_bytes,
        "/usr/lib/libSystem.B.dylib".to_string(),
        "very_very_very_very_very_very_very_longid".to_string(),
    );

    let changed_macho = MachO::parse(&changed, 0).unwrap();

    insta::assert_debug_snapshot!(changed_macho);
    insta::assert_debug_snapshot!(changed_macho.libs);
    insta::assert_snapshot!(changed.len(), @"33440");
}
