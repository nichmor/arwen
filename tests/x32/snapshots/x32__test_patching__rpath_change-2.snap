---
source: tests/x32/test_patching.rs
expression: changed_macho.rpaths
---
[
    "path_graf_path_graf_path_graf_path_graf",
]
